### PR TITLE
Make minmdns use a distinct policy for endpoint and IP choices

### PR DESCRIPTION
**Pull request by andy31415**
**Reason for 1.0**
   - provides interop hooks (we observed without such hooks, some linux environments are problematic, specifically for TC_RR)
   - provides significant usability improvements on platforms which have a strong opinion on where/how minmdns should advertise.

#### Problem
MinMDNS attempts to listen and advertise **all** endpoints and IP addresses on a device.

This is not always appropriate, one example being linux which supports `deprecated` and `temporary` IPv6 addresses, which should generally not be reported/used.

#### Change overview

This is part of preparation for #22046 - either core SDK or app-integration logic should allow for defining distinct MinMDNS policies for endpoint and ip selection.

This splits the endpoint and address selection into a separate library which is compile-time enabled, with the expectation that we can build an alternate policy (e.g. for linux using libnl for more info on ip addresses) or let the app integration decide to link their own library policy.

#### Testing
Manually checked with minmdns client/server examples.
CI will validate as linux integration tests rely on minmdns.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/lib/dnssd/minimal_mdns/responders/IP.cpp [hunk 1]
```diff
@@ -16,18 +16,24 @@
...
 {
     chip::Inet::IPAddress addr;
-    for (chip::Inet::InterfaceAddressIterator it; it.HasCurrent(); it.Next())
+    UniquePtr<IpAddressIterator> ips = Policies::GetIpAddressesForEndpoint(source->Interface);
+    VerifyOrDie(ips);
+
+    while (ips->Next(addr))
     {
-        if ((it.GetInterfaceId() == source->Interface) && (it.GetAddress(addr) == CHIP_NO_ERROR) && addr.IsIPv4())
+        if (addr.IsIPv4())
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/minimal_mdns/responders/IP.cpp:36]: This seems costly to iterate over all addresses twice, rather than having a separate method to interate over IPv4 addresses and IPv6 addresses, and where someone could decide to not implement IPv4? They could still use the same logic under the hood and just filter, if that works well for them, but if it works fine to iterate separately for a platform, that could be accomodated as well.

File: src/lib/dnssd/minimal_mdns/responders/IP.cpp [hunk 2]
```diff
@@ -16,18 +16,24 @@
...
 
+using chip::Platform::UniquePtr;
+
 void IPv4Responder::AddAllResponses(const chip::Inet::IPPacketInfo * source, ResponderDelegate * delegate,
                                     const ResponseConfiguration & configuration)
 {
     chip::Inet::IPAddress addr;
-    for (chip::Inet::InterfaceAddressIterator it; it.HasCurrent(); it.Next())
+    UniquePtr<IpAddressIterator> ips = Policies::GetIpAddressesForEndpoint(source->Interface);
+    VerifyOrDie(ips);
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/minimal_mdns/responders/IP.cpp:32]: Please just early exit on failure to alloc. Crashing is quite harsh

File: src/lib/dnssd/minimal_mdns/BUILD.gn [hunk 3]
```diff
@@ -50,6 +53,22 @@ config("config") {
...
+}
+
+static_library("default_policy") {
+  sources = [ "AddressPolicy_DefaultImpl.cpp" ]
+
+  public_deps = [
+    ":policy_header",
+    "${chip_root}/src/inet",
+  ]
+}
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/minimal_mdns/BUILD.gn:70]: This seems to require the source be built  at the same time here, from within the chip checkout, even though the iterator is an interface.

It would be much more flexible to just define the interface in the header, but allow the implementers to pass/initialize the factory or iterator-getter in the initialization of minimal mdns server. That would also allow more easily configuring the object prior to its use, rather than expecting what amounts to a singular context-free global iterator

Comment (andy31415) [reply src/lib/dnssd/minimal_mdns/BUILD.gn:70]: Will update after side conversation about layout.

TLDR: will move policy as an abstract base class that can be injected.
Builds will be able to even exclude the default policy, to make sure their custom policy is actually enforced/used.

Comment (andy31415) [reply src/lib/dnssd/minimal_mdns/BUILD.gn]: Updated with abstract base class.

File: src/lib/dnssd/minimal_mdns/responders/IP.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/dnssd/minimal_mdns/responders/IP.cpp:32]: Returning no IP addresses seems very hard to debug. Are you sure it is a better way of working?

Comment (andy31415) [reply src/lib/dnssd/minimal_mdns/responders/IP.cpp:36]: So thinking of options:
  - apply `ostrich approach/head in sand` and just drop it. If we run out of mem (not as likely as these objects are basically a vtable) then we will crash on vtable access Next(). I think we do that in other places.
  - This explicit crash as we have now (ugly to crash, but explicit crash location at least).
  - if condition with nice log of 'memory allocation error'. Seemed overkill.

**Review by bzbarsky-apple — COMMENTED** [review]
@andy31415 Should some of this policy infrastructure be shared by the group multicast bits too?

**Review by tcarmelveilleux — APPROVED** [review]
